GitUI/CommandsDialogs/FormCommit.cs

```diff
@@ -2199,12 +2199,12 @@ private void SetCommitMessageFromTextBox(string commitMessageText)
                 {
                     // When a committemplate is used, skip comments
                     // otherwise: "#" is probably not used for comment but for issue number
-                    if ((!string.IsNullOrEmpty(line) && !line.StartsWith("#")) ||
+                    if ((line != null && !line.StartsWith("#")) ||
                         string.IsNullOrEmpty(_commitTemplate))
```

Comment:
If `string.IsNullOrEmpty(_commitTemplate)` is true, this lets the (very unlikely) `line == null` into the block to the final `WriteLine(line)`.
The logic was and is quite hard to read. And `AppSettings.AddNewlineToCommitMessageWhenMissing` is not self-explaining: In the settings dialog, it is labeled `"Ensure the second line of commit message is empty"`.
As @RussKie requested, we should really have unit tests for this function.
Before I try to describe everything (positive logic, combine nested `if`, etc.), I take the liberty to suggest:
```C#
            using (var textWriter = new StreamWriter(path, false, encoding))
            {
                bool addNewlineToCommitMessageWhenMissing = AppSettings.AddNewlineToCommitMessageWhenMissing;
                bool usingCommitTemplate = !string.IsNullOrEmpty(_commitTemplate);

                var lineNumber = 1;
                foreach (var line in commitMessageText.Split('\n'))
                {
                    string nonNullLine = line ?? string.Empty;

                    // When a committemplate is used, skip comments and do not count them as line.
                    // otherwise: "#" is probably not used for comment but for issue number
                    if (usingCommitTemplate && nonNullLine.StartsWith("#"))
                    {
                        continue;
                    }

                    if (addNewlineToCommitMessageWhenMissing && lineNumber == 2 && !string.IsNullOrEmpty(nonNullLine))
                    {
                        textWriter.WriteLine();
                    }

                    textWriter.WriteLine(nonNullLine);

                    lineNumber++;
                }
            }
```

Reply:
> And `AppSettings.AddNewlineToCommitMessageWhenMissing` is not self-explaining: In the settings dialog, it is labeled `"Ensure the second line of commit message is empty"`.

I've noticed that too.
We should rename it, but it can be done separately. Perhaps call it `EnsureCommitSecondLineEmpty`?